src/wasm_fetcher/mod.rs

```diff
@@ -28,8 +30,8 @@ pub(crate) fn parse_wasm_url(
 
   match parsed_url.scheme() {
     "file" => Ok(Box::new(Local::new(parsed_url.path())?)),
-    "http" => Ok(Box::new(Https::new(url.parse::<Uri>()?, remote_insecure)?)),
-    "https" => Ok(Box::new(Https::new(url.parse::<Uri>()?, remote_insecure)?)),
+    "http" | "https" => Ok(Box::new(Https::new(url.parse::<Url>()?, remote_insecure)?)),
+    "registry" => Ok(Box::new(Registry::new(parsed_url)?)),
```

Comment:
The registry handler must take these flags at creation time:
  * `remote_insecure`: the remote OCI registry is secured with a self-signed tls key
  * `remote_non_tls`: the remote OCI registry is not secured by tls, it's served over plain http

Reply:
I don't see a straightforward way to configure this so that `oci-distribution` honors these settings. On their `ClientConfig` struct there's only the `protocol` that can be configured, but that only signals whether you want to use Http, Https, or Https except for a fixed number of hosts that will use Http.

Besides this, we probably want to check `remote_non_tls` again because we are setting it, but no backend is honoring it (the HTTP one is not honoring it either).

Reply:
PR updated so it honors the `remote_non_tls` setting. I don't find a way to honor `remote_insecure` though (HTTPS but ignoring the presented certificate).

As a second question, given that the user can provide `--wasm-uri` with `http`, `https` and `registry` I specially wonder about how `http` and `https` interact from the UX perspective with `wasm-remote-non-tls` specially and `wasm-remote-insecure` as well, given that they are somehow tied to what schema the user provides on `--wasm-uri`.

Reply:
> Besides this, we probably want to check remote_non_tls again because we are setting it, but no backend is honoring it (the HTTP one is not honoring it either).

The `https` fetcher handles the `remote_non_tls` parameter automatically. If the URL scheme is `http://`, it will pull from a non-tls terminated server.

Reply:
> PR updated so it honors the remote_non_tls setting. I don't find a way to honor remote_insecure though (HTTPS but ignoring the presented certificate).

I fear you will have to look into the OCI library and see which http client it uses. Based on that you will have to do some dancing around to make it ignore the remote tls certificate when this is not valid. I suspect this will end up with an upstream contribution

Reply:
> As a second question, given that the user can provide --wasm-uri with file, http, https and registry I specially wonder about how http and https interact from the UX perspective with wasm-remote-non-tls specially and wasm-remote-insecure as well, given that they are somehow tied to what schema the user provides on --wasm-uri.

Let's park that for a short while, I think we will have to drop these flags once we move to support multiple WASM policies per server (see https://github.com/chimera-kube/policy-server/issues/5).